File: backend/services/gmailService.js
// backend/services/gmailService.js

const { google } = require('googleapis');
require('dotenv').config();

// ─── CREATE OAUTH CLIENT ─────────────────────────────

// This creates the Google authentication client
// Used to generate login URLs and exchange codes for tokens
const createOAuth2Client = () => {
  return new google.auth.OAuth2(
    process.env.GOOGLE_CLIENT_ID,
    process.env.GOOGLE_CLIENT_SECRET,
    // This URL is where Google sends user after they approve
    // Must exactly match what you set in Google Cloud Console
    `${process.env.BACKEND_URL || 'http://localhost:5000'}/api/gmail/callback`
  );
};

// ─── GENERATE AUTH URL ───────────────────────────────

// Creates the Google login URL that user clicks
// userId is passed as state so we know which user
// connected Gmail after they come back from Google
const getAuthUrl = (userId) => {
  const oauth2Client = createOAuth2Client();
  
  return oauth2Client.generateAuthUrl({
    access_type: 'offline',  // Gets refresh token for long term access
    prompt: 'consent',       // Always show consent screen
    scope: [
      // This is the ONLY permission we request
      // Read only — we can never send or delete emails
      'https://www.googleapis.com/auth/gmail.readonly'
    ],
    state: userId  // Passed back to callback so we know who this is
  });
};

// ─── GET TOKENS FROM CODE ────────────────────────────

// After user approves Google sends a code
// This exchanges that code for actual access tokens
const getTokensFromCode = async (code) => {
  const oauth2Client = createOAuth2Client();
  const { tokens } = await oauth2Client.getToken(code);
  return tokens;
};

// ─── CREATE GMAIL CLIENT ─────────────────────────────

// Creates an authenticated Gmail API client
// Used to actually read emails
const createGmailClient = (accessToken, refreshToken) => {
  const oauth2Client = createOAuth2Client();
  
  // Set the user's tokens
  oauth2Client.setCredentials({
    access_token: accessToken,
    refresh_token: refreshToken
  });
  
  // Return gmail client ready to use
  return google.gmail({ version: 'v1', auth: oauth2Client });
};

// ─── SEARCH RECEIPT EMAILS ───────────────────────────

// Searches Gmail for emails that look like receipts
// Uses two search passes to catch as many subscription emails as possible
const searchReceiptEmails = async (gmail) => {
  
  console.log('Searching Gmail with multiple queries...')

  // Query 1: Payment/billing keywords (searches full email)
  const paymentQuery = '(receipt OR invoice OR payment OR subscription OR renewal OR billing OR charged OR membership OR "payment method" OR "next billing" OR "successfully processed" OR "amount paid" OR "plan renewed") newer_than:1y'

  // Query 2: Known service names in subject (catches test emails)
  const serviceQuery = 'subject:(netflix OR spotify OR chatgpt OR canva OR "youtube premium" OR "amazon prime" OR hotstar OR adobe OR notion OR github OR linkedin OR zoom OR "google one" OR icloud OR nordvpn) newer_than:1y'

  const [paymentRes, serviceRes] = await Promise.all([
    gmail.users.messages.list({ userId: 'me', q: paymentQuery, maxResults: 200 }),
    gmail.users.messages.list({ userId: 'me', q: serviceQuery, maxResults: 50 })
  ])

  const paymentMessages = paymentRes.data.messages || []
  const serviceMessages = serviceRes.data.messages || []

  // Deduplicate by message ID
  const seen = new Set()
  const allMessages = []

  for (const msg of [...paymentMessages, ...serviceMessages]) {
    if (!seen.has(msg.id)) {
      seen.add(msg.id)
      allMessages.push(msg)
    }
  }

  console.log(`Found ${paymentMessages.length} payment emails + ${serviceMessages.length} service emails = ${allMessages.length} unique`)

  return allMessages
}

const decodeBase64Url = (data) => {
  if (!data) return '';
  try {
    return Buffer.from(data, 'base64url').toString('utf8');
  } catch (err) {
    try {
      return Buffer.from(data, 'base64').toString('utf8');
    } catch (e) {
      return '';
    }
  }
};

const findTextPart = (part) => {
  if (!part) return null;
  if (part.mimeType === 'text/plain' && part.body && part.body.data) {
    return decodeBase64Url(part.body.data);
  }
  if (part.parts) {
    for (const subPart of part.parts) {
      const found = findTextPart(subPart);
      if (found) return found;
    }
  }
  return null;
};

const findHtmlPart = (part) => {
  if (!part) return null;
  if (part.mimeType === 'text/html' && part.body && part.body.data) {
    const html = decodeBase64Url(part.body.data);
    let clean = html;
    
    // Fast non-backtracking style removal
    while (clean.includes('<style') && clean.includes('</style>')) {
      const start = clean.indexOf('<style');
      const end = clean.indexOf('</style>', start);
      if (end === -1) break;
      clean = clean.substring(0, start) + clean.substring(end + 8);
    }
    
    // Fast non-backtracking script removal
    while (clean.includes('<script') && clean.includes('</script>')) {
      const start = clean.indexOf('<script');
      const end = clean.indexOf('</script>', start);
      if (end === -1) break;
      clean = clean.substring(0, start) + clean.substring(end + 9);
    }
    
    return clean
      .replace(/<[^>]*>/g, ' ')
      .replace(/\s+/g, ' ')
      .trim();
  }
  if (part.parts) {
    for (const subPart of part.parts) {
      const found = findHtmlPart(subPart);
      if (found) return found;
    }
  }
  return null;
};

const getBodyText = (payload) => {
  if (!payload) return '';
  if (payload.body && payload.body.data) {
    const text = decodeBase64Url(payload.body.data);
    if (payload.mimeType === 'text/html') {
      let clean = text;
      while (clean.includes('<style') && clean.includes('</style>')) {
        const start = clean.indexOf('<style');
        const end = clean.indexOf('</style>', start);
        if (end === -1) break;
        clean = clean.substring(0, start) + clean.substring(end + 8);
      }
      while (clean.includes('<script') && clean.includes('</script>')) {
        const start = clean.indexOf('<script');
        const end = clean.indexOf('</script>', start);
        if (end === -1) break;
        clean = clean.substring(0, start) + clean.substring(end + 9);
      }
      return clean
        .replace(/<[^>]*>/g, ' ')
        .replace(/\s+/g, ' ')
        .trim();
    }
    return text;
  }
  let body = findTextPart(payload);
  if (!body) {
    body = findHtmlPart(payload);
  }
  return body || '';
};

// Gets the actual content of a specific email
// Uses the proper getBodyText helper for correct base64url decoding
const getEmailContent = async (gmail, messageId) => {
  const response = await gmail.users.messages.get({
    userId: 'me',
    id: messageId,
    format: 'full'
  })

  const headers = response.data.payload.headers
  const subject = headers.find(h => h.name === 'Subject')?.value || ''
  const from = headers.find(h => h.name === 'From')?.value || ''
  const date = headers.find(h => h.name === 'Date')?.value || ''
  const snippet = response.data.snippet || ''

  // Extract body using the proper helper that handles
  // base64url decoding, HTML stripping, and nested MIME parts
  const fullBody = getBodyText(response.data.payload)

  // Clean and truncate for AI processing
  const cleanBody = fullBody
    .replace(/&nbsp;/g, ' ')
    .replace(/&amp;/g, '&')
    .replace(/\s+/g, ' ')
    .trim()
    .substring(0, 3000)

  return {
    subject,
    from,
    date,
    snippet,
    body: cleanBody,
    messageId
  }
};

// Export all functions
module.exports = {
  getAuthUrl,
  getTokensFromCode,
  createGmailClient,
  searchReceiptEmails,
  getEmailContent
};

File: backend/services/gmail_service.py
# backend/services/gmail_service.py
# Gmail API service — exact port of services/gmailService.js

import re
import base64
from google.oauth2.credentials import Credentials
from googleapiclient.discovery import build
from config.settings import get_settings

settings = get_settings()


# ─── CREATE GMAIL CLIENT ─────────────────────────────

def create_gmail_client(access_token: str, refresh_token: str):
    """Creates an authenticated Gmail API client — mirrors createGmailClient()"""
    credentials = Credentials(
        token=access_token,
        refresh_token=refresh_token,
        token_uri="https://oauth2.googleapis.com/token",
        client_id=settings.GOOGLE_CLIENT_ID,
        client_secret=settings.GOOGLE_CLIENT_SECRET
    )
    return build("gmail", "v1", credentials=credentials)


# ─── GENERATE AUTH URL ────────────────────────────────

def get_auth_url(user_id: str) -> str:
    """Creates the Google OAuth login URL — mirrors getAuthUrl()"""
    from google_auth_oauthlib.flow import Flow

    flow = Flow.from_client_config(
        {
            "web": {
                "client_id": settings.GOOGLE_CLIENT_ID,
                "client_secret": settings.GOOGLE_CLIENT_SECRET,
                "auth_uri": "https://accounts.google.com/o/oauth2/auth",
                "token_uri": "https://oauth2.googleapis.com/token",
                "redirect_uris": [f"{settings.BACKEND_URL}/api/gmail/callback"]
            }
        },
        scopes=["https://www.googleapis.com/auth/gmail.readonly"]
    )
    flow.redirect_uri = f"{settings.BACKEND_URL}/api/gmail/callback"

    auth_url, _ = flow.authorization_url(
        access_type="offline",
        prompt="consent",
        state=user_id,
        include_granted_scopes="true"
    )
    return auth_url


# ─── GET TOKENS FROM CODE ────────────────────────────

def get_tokens_from_code(code: str) -> dict:
    """Exchanges authorization code for tokens — mirrors getTokensFromCode()"""
    from google_auth_oauthlib.flow import Flow

    flow = Flow.from_client_config(
        {
            "web": {
                "client_id": settings.GOOGLE_CLIENT_ID,
                "client_secret": settings.GOOGLE_CLIENT_SECRET,
                "auth_uri": "https://accounts.google.com/o/oauth2/auth",
                "token_uri": "https://oauth2.googleapis.com/token",
                "redirect_uris": [f"{settings.BACKEND_URL}/api/gmail/callback"]
            }
        },
        scopes=["https://www.googleapis.com/auth/gmail.readonly"]
    )
    flow.redirect_uri = f"{settings.BACKEND_URL}/api/gmail/callback"
    flow.fetch_token(code=code)

    creds = flow.credentials
    return {
        "access_token": creds.token,
        "refresh_token": creds.refresh_token
    }


# ─── SEARCH RECEIPT EMAILS ───────────────────────────

def search_receipt_emails(gmail) -> list:
    """Searches Gmail for receipt/billing emails — mirrors searchReceiptEmails()"""
    print("Searching Gmail with multiple queries...")

    # Query 1: Payment/billing keywords
    payment_query = (
        '(receipt OR invoice OR payment OR subscription OR renewal OR billing '
        'OR charged OR membership OR "payment method" OR "next billing" '
        'OR "successfully processed" OR "amount paid" OR "plan renewed") newer_than:1y'
    )

    # Query 2: Known service names in subject
    service_query = (
        'subject:(netflix OR spotify OR chatgpt OR canva OR "youtube premium" '
        'OR "amazon prime" OR hotstar OR adobe OR notion OR github OR linkedin '
        'OR zoom OR "google one" OR icloud OR nordvpn) newer_than:1y'
    )

    payment_res = gmail.users().messages().list(
        userId="me", q=payment_query, maxResults=200
    ).execute()

    service_res = gmail.users().messages().list(
        userId="me", q=service_query, maxResults=50
    ).execute()

    payment_messages = payment_res.get("messages", [])
    service_messages = service_res.get("messages", [])

    # Deduplicate by message ID
    seen = set()
    all_messages = []
    for msg in payment_messages + service_messages:
        if msg["id"] not in seen:
            seen.add(msg["id"])
            all_messages.append(msg)

    print(f"Found {len(payment_messages)} payment emails + {len(service_messages)} service emails = {len(all_messages)} unique")
    return all_messages


# ─── DECODE BASE64URL ─────────────────────────────────

def decode_base64url(data: str) -> str:
    """Decodes base64url encoded data — mirrors decodeBase64Url()"""
    if not data:
        return ""
    try:
        # Add padding if needed
        padded = data + "=" * (4 - len(data) % 4) if len(data) % 4 else data
        return base64.urlsafe_b64decode(padded).decode("utf-8", errors="replace")
    except Exception:
        return ""


# ─── FIND TEXT/HTML PARTS ─────────────────────────────

def find_text_part(part: dict) -> str | None:
    """Recursively finds text/plain part — mirrors findTextPart()"""
    if not part:
        return None
    if part.get("mimeType") == "text/plain" and part.get("body", {}).get("data"):
        return decode_base64url(part["body"]["data"])
    for sub_part in part.get("parts", []):
        found = find_text_part(sub_part)
        if found:
            return found
    return None


def find_html_part(part: dict) -> str | None:
    """Recursively finds text/html part and strips HTML — mirrors findHtmlPart()"""
    if not part:
        return None
    if part.get("mimeType") == "text/html" and part.get("body", {}).get("data"):
        html = decode_base64url(part["body"]["data"])
        clean = html

        # Remove style tags
        while "<style" in clean and "</style>" in clean:
            start = clean.index("<style")
            end = clean.index("</style>", start)
            if end == -1:
                break
            clean = clean[:start] + clean[end + 8:]

        # Remove script tags
        while "<script" in clean and "</script>" in clean:
            start = clean.index("<script")
            end = clean.index("</script>", start)
            if end == -1:
                break
            clean = clean[:start] + clean[end + 9:]

        # Strip remaining HTML tags
        clean = re.sub(r"<[^>]*>", " ", clean)
        clean = re.sub(r"\s+", " ", clean).strip()
        return clean

    for sub_part in part.get("parts", []):
        found = find_html_part(sub_part)
        if found:
            return found
    return None


def get_body_text(payload: dict) -> str:
    """Extracts body text from email payload — mirrors getBodyText()"""
    if not payload:
        return ""

    if payload.get("body", {}).get("data"):
        text = decode_base64url(payload["body"]["data"])
        if payload.get("mimeType") == "text/html":
            clean = text
            while "<style" in clean and "</style>" in clean:
                start = clean.index("<style")
                end = clean.index("</style>", start)
                if end == -1:
                    break
                clean = clean[:start] + clean[end + 8:]
            while "<script" in clean and "</script>" in clean:
                start = clean.index("<script")
                end = clean.index("</script>", start)
                if end == -1:
                    break
                clean = clean[:start] + clean[end + 9:]
            clean = re.sub(r"<[^>]*>", " ", clean)
            clean = re.sub(r"\s+", " ", clean).strip()
            return clean
        return text

    body = find_text_part(payload)
    if not body:
        body = find_html_part(payload)
    return body or ""


# ─── GET EMAIL CONTENT ────────────────────────────────

def get_email_content(gmail, message_id: str) -> dict:
    """Gets full content of a specific email — mirrors getEmailContent()"""
    response = gmail.users().messages().get(
        userId="me", id=message_id, format="full"
    ).execute()

    headers = response.get("payload", {}).get("headers", [])
    subject = next((h["value"] for h in headers if h["name"] == "Subject"), "")
    from_addr = next((h["value"] for h in headers if h["name"] == "From"), "")
    date = next((h["value"] for h in headers if h["name"] == "Date"), "")
    snippet = response.get("snippet", "")

    # Extract body using helper
    full_body = get_body_text(response.get("payload", {}))

    # Clean and truncate for AI processing
    clean_body = full_body \
        .replace("&nbsp;", " ") \
        .replace("&amp;", "&")
    clean_body = re.sub(r"\s+", " ", clean_body).strip()[:3000]

    return {
        "subject": subject,
        "from": from_addr,
        "date": date,
        "snippet": snippet,
        "body": clean_body,
        "messageId": message_id
    }
